File: packages/provide-uterm-server/src/provide/uterm/server/bridge/hub/registry.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""WorkerRegistry: in-memory registry of attached workers.

Owns the lifecycle of :class:`WorkerTermState` instances keyed by
``worker_id``. This class is intentionally a thin wrapper around a
``dict`` — its purpose is to give the worker map a *name* that other
hub services can hold a reference to, instead of poking at
``TermHub._workers`` directly.

Lock semantics are unchanged from the pre-refactor design: lock-free
reads remain safe because CPython dict reads are atomic, and mutations
are coordinated by callers via :attr:`TermHub._lock`. The registry
itself takes no locks.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterator

    from provide.uterm.server.bridge.models import WorkerTermState


class WorkerRegistry:
    """In-memory registry of attached workers, keyed by ``worker_id``.

    The underlying ``dict`` is exposed as :attr:`_workers` so existing
    hub mixin code can continue to use mapping operations
    (``setdefault``, ``items``, ``values``, indexed access, ``del``)
    directly during the phased refactor. New code should prefer the
    explicit methods (:meth:`get`, :meth:`put`, :meth:`pop`,
    :meth:`contains`, :meth:`all`, :meth:`items`).
    """

    __slots__ = ("_workers",)

    def __init__(self) -> None:
        self._workers: dict[str, WorkerTermState] = {}

    # -- Explicit accessors (preferred for new code) -----------------------

    def get(self, worker_id: str) -> WorkerTermState | None:
        """Return the state for *worker_id* or ``None`` if unknown."""
        return self._workers.get(worker_id)

    def require(self, worker_id: str) -> WorkerTermState:
        """Return the state for *worker_id* or raise ``KeyError``."""
        st = self._workers.get(worker_id)
        if st is None:
            raise KeyError(worker_id)
        return st

    def put(self, worker_id: str, state: WorkerTermState) -> None:
        """Insert or replace the state for *worker_id*."""
        self._workers[worker_id] = state

    def setdefault(self, worker_id: str, state: WorkerTermState) -> WorkerTermState:
        """Return the existing state for *worker_id*, or insert *state* and return it."""
        return self._workers.setdefault(worker_id, state)

    def pop(self, worker_id: str) -> WorkerTermState | None:
        """Remove and return the state for *worker_id*, or ``None`` if absent."""
        return self._workers.pop(worker_id, None)

    def discard(self, worker_id: str) -> bool:
        """Remove *worker_id* if present; return True if removed."""
        return self._workers.pop(worker_id, None) is not None

    def contains(self, worker_id: str) -> bool:
        """Return True if *worker_id* is registered."""
        return worker_id in self._workers

    def all(self) -> list[WorkerTermState]:
        """Return a snapshot list of all registered worker states."""
        return list(self._workers.values())

    def keys(self) -> list[str]:
        """Return a snapshot list of all registered worker ids."""
        return list(self._workers.keys())

    def items(self) -> list[tuple[str, WorkerTermState]]:
        """Return a snapshot list of (worker_id, state) tuples."""
        return list(self._workers.items())

    # -- Dunder convenience ------------------------------------------------

    def __len__(self) -> int:
        return len(self._workers)

    def __iter__(self) -> Iterator[str]:
        return iter(self._workers)

    def __contains__(self, worker_id: object) -> bool:
        return worker_id in self._workers


__all__ = ["WorkerRegistry"]

File: packages/provide-uterm-ts/src/hub/registry.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * In-memory registry of attached workers, keyed by worker id.
 *
 * A deliberately thin wrapper over a `Map`: its job is to give the worker
 * table a name that the hub's other services reference, and to make every
 * read return a *snapshot*.
 *
 * That snapshot rule is the substance. Callers iterate the worker set while
 * the hub attaches and detaches workers, so handing out a live view would
 * mean iteration observing a mutation — skipped entries at best, a thrown
 * iterator at worst. `keys`, `all` and `items` therefore copy.
 *
 * The registry takes no locks. Each method is a single map operation, so
 * there is no window inside one for another caller to observe; higher-level
 * invariants that span several operations are the hub's to coordinate.
 *
 * Port of the Python module `provide.uterm.server.bridge.hub.registry` and
 * the Go package `hub`.
 */

/** Registry of worker states, keyed by worker id. */
export class WorkerRegistry<T> {
  readonly #workers = new Map<string, T>();

  /** How many workers are registered. */
  get size(): number {
    return this.#workers.size;
  }

  /** The state for `workerId`, or `undefined` when it is unknown. */
  get(workerId: string): T | undefined {
    return this.#workers.get(workerId);
  }

  /**
   * The state for `workerId`.
   *
   * @throws {Error} When the worker is unknown, for call sites where an
   *   absent worker is a bug rather than a case to handle.
   */
  require(workerId: string): T {
    const state = this.#workers.get(workerId);
    if (state === undefined) {
      throw new Error(workerId);
    }
    return state;
  }

  /** Insert or replace the state for `workerId`. */
  put(workerId: string, state: T): void {
    this.#workers.set(workerId, state);
  }

  /**
   * The existing state for `workerId`, inserting `state` if there is none.
   *
   * The existing entry wins, so an attach racing a reattach cannot discard
   * live state and hand back a fresh one.
   */
  setDefault(workerId: string, state: T): T {
    const existing = this.#workers.get(workerId);
    if (existing !== undefined) {
      return existing;
    }
    this.#workers.set(workerId, state);
    return state;
  }

  /** Remove and return the state for `workerId`, if it was present. */
  pop(workerId: string): T | undefined {
    const state = this.#workers.get(workerId);
    this.#workers.delete(workerId);
    return state;
  }

  /** Remove `workerId`, reporting whether anything was removed. */
  discard(workerId: string): boolean {
    return this.#workers.delete(workerId);
  }

  /** Whether `workerId` is registered. */
  contains(workerId: string): boolean {
    return this.#workers.has(workerId);
  }

  /** A snapshot of every registered state, in insertion order. */
  all(): T[] {
    return [...this.#workers.values()];
  }

  /** A snapshot of every registered worker id, in insertion order. */
  keys(): string[] {
    return [...this.#workers.keys()];
  }

  /** A snapshot of every (worker id, state) pair, in insertion order. */
  items(): Array<[string, T]> {
    return [...this.#workers.entries()];
  }

  /** Iterate the registered worker ids. */
  [Symbol.iterator](): Iterator<string> {
    return this.#workers.keys();
  }
}
